ruby/phase-1.md

```diff
@@ -0,0 +1,616 @@
+      activity.logger.info('❤️')
+      activity.logger.debug("Attempt #{activity.info.attempts}")
+      activity.heartbeat('details')
+    end
+  rescue Temporal::Error::CancellationRequested
+    activity.logger.info('💀')
+    # Accept cancellation request
+    raise Temporal::Error::ActivityCancelled, 'cancellation details here'
+  end
+end
+```
+
+Notes:
+
+- A custom activity name can be specified using `activity 'my-activity'` call (we can't use name,
+  because it is reserved to the actual class name in Ruby)
+- `activity` within `#execute` is an instance of `Temporal::Activity::Context`
+- `activity.logger` will use a logging adapter configure via `Temporal::Configuration` (more on
+  this in the 2nd phase)
+- Common logging context can be configured by using middleware (more on this in the 2nd phase)
+- `activity.heartbeat` is a synchronous call
+- In case of a cancellation request `Temporal::Error::CancellationRequested` will get raised, which
+  can be accepted by raising `Temporal::Error::ActivityCancelled`
+
+```ruby
+class Temporal::Activity::Context
+  # Access to a pre-configured logger
+  def logger() -> Temporal::Logger
+
+  # Heartbeat will raise if a cancellation was requested
+  def heartbeat(*details) -> void
+
+  # Provides access to extra details from PollActivityTaskQueueResponse
+  def info() -> Temporal::Info::Activity
+end
+```
+
+### Return values and errors
+
+An activity can communicate its result to the workflow by returning a value or raising an error. The
+return values will get serialized as a `Payload` object similar to activity input arguments.
+Similarly, an error will get serialized and re-raised from within the workflow (if an activity
+called "synchronously").
+
+The main idea here is to make it feel (from the workflow's perspective) like a local invocation of
+an activity without any additional constructs in the way.
+
+```ruby
+class MakePaymentActivity < Temporal::Activity
+  PaymentMethodNotFound = Class.new(Temporal::Error::ActivityFailure)
+  PaymentFailed = Class.new(Temporal::Error::ActivityFailure)
+
+  def execute(payment_method_id)
+    payment_method = PaymentMethod.find(payment_method_id)
+    raise UserNotFound, "No payment_method with id #{payment_method_id}" unless payment_method
+
+    result = PaymentService.execute(user.payment_method)
+    raise PaymentFailed, "Unable to process payment: #{result.error}" if result.failure?
+
+    result.payment_id
+  end
+end
+```
+
+Notes:
+
+- Errors defined within the activity class are namespaced as `MyActivity::PaymentFailed`,
+  which is a very useful property
+  - This will however not work if a workflow does not have access to the activity definition
+  - In this case it will get raised as a generic `Temporal::Error::ActivityFailure` with an original
+    error's class name available via `#original_error`
+- Subclassing `Temporal::Error::ActivityFailure` allows us to inject extra information into the
+  error (such as activity name, id, etc)
```

Comment:
ActivityFailure should not be subclassed. It should never be thrown by the application code. Any application level error should be returned as ActivityFailure that has the ApplicationFailure as a cause.

Reply:
I didn't mean to reference Proto's ActivityFailure, it's just a coincidence. It's going to be an SDK-specific error base-class.